Stop exporting TH/THC symbols as extern "C".

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29919 Stop exporting TH/THC symbols as extern "C".**

This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)
Bug reference: https://github.com/RadeonOpenCompute/hcc/issues/839

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Stop exporting TH/THC symbols as extern "C".

This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit ed87428:
* **9/9** failures introduced in this PR
* **0/9** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=ed8742839740655aa693061138a54557412bc397):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_ios_10_2_1_x86_64_build | Build | `/Users/distiller/project/aten/src/TH/THLogAdd.h:8:21: error:` |
| pytorch_ios_10_2_1_arm64_build | Build | `/Users/distiller/project/aten/src/TH/THLogAdd.h:8:21: error:` |
| binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build | Build | `/pytorch/aten/src/TH/THLogAdd.h:8:21: error:` |
| caffe2_onnx_py2_gcc5_ubuntu16_04_build | Build | `/var/lib/jenkins/workspace/aten/src/TH/THLogAdd.h:8:21: error:` |
| pytorch_linux_xenial_py2_7_9_build | Build | `/var/lib/jenkins/workspace/aten/src/TH/THLogAdd.h:8:21: error:` |
| binary_linux_manywheel_2_7mu_cpu_devtoolset7_build | Build | `/pytorch/aten/src/TH/THLogAdd.h:8:21: error:` |
| caffe2_onnx_py3_6_clang7_ubuntu16_04_build | Build | `/var/lib/jenkins/workspace/aten/src/TH/THLogAdd.h:8:21: error:` |
| pytorch_linux_xenial_py3_clang5_asan_build | Build | `/var/lib/jenkins/workspace/aten/src/TH/THLogAdd.h:8:21: error:` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Stop exporting TH/THC symbols as extern "C"."


This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)
Bug reference: https://github.com/RadeonOpenCompute/hcc/issues/839

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop exporting TH/THC symbols as extern "C"."


This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)
Bug reference: https://github.com/RadeonOpenCompute/hcc/issues/839

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

iotamudelta reviewed (approved):
Thanks for working around this!

ezyang committed:
Update on "Stop exporting TH/THC symbols as extern "C"."


This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)
Bug reference: https://github.com/RadeonOpenCompute/hcc/issues/839

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Stop exporting TH/THC symbols as extern "C"."


This is back from the old days when we had C clients, but we no
longer have this.  Switch off extern "C".

This makes us less likely to run into a ROCm bug where functions
declared __host__ do not respect extern "C" in the header forward
declaration (you have to explicitly respecify it on the implementation.)
Bug reference: https://github.com/RadeonOpenCompute/hcc/issues/839

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

gchanan commented:
@ezyang did you still want a review of this?

ezyang commented:
I think this patch is dead; also @anjali411 redid some of it in some recent complex stuff.

anjali411 commented:
I removed `TH_EXTERNC` in this [PR](https://github.com/pytorch/pytorch/pull/36142).  However, `THC_EXTERNC` and `TH_CPP_API` still need to removed
